File: 3_backtracking/3_Word_Search_II.cpp
// Word Search II

/*
Given a 2D board and a list of words from the dictionary, find all words in the board.

Each word must be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.

For example,
Given words = ["oath","pea","eat","rain"] and board =

[
  ['o','a','a','n'],
  ['e','t','a','e'],
  ['i','h','k','r'],
  ['i','f','l','v']
]
Return ["eat","oath"].
Note:
You may assume that all inputs are consist of lowercase letters a-z.

click to show hint.

You would need to optimize your backtracking to pass the larger test. Could you stop backtracking earlier?

If the current candidate does not exist in all words' prefix, you could stop backtracking immediately. What kind of data structure could answer such query efficiently? Does a hash table work? Why or why not? How about a Trie? If you would like to learn how to implement a basic trie, please work on this problem: Implement Trie (Prefix Tree) first.
*/

// my solution
class TrieNode {
public:
    bool is_end;
    vector<TrieNode*> children;
    
    TrieNode() {
        is_end = false;
        children = vector<TrieNode*>(26, NULL);
    }
};

class Trie {
public:
    TrieNode* getRoot() {return root;}
    Trie(vector<string>& words) {
        root = new TrieNode();
        for (int i = 0; i < words.size(); ++i)
            addWord(words[i]);
    }
    
    void addWord(const string& word) {
        TrieNode* cur = root;
        
        for (int i = 0; i < word.size(); i++) {
            int idx = word[i] - 'a';
            
            if (cur->children[idx] == NULL)
                cur->children[idx] = new TrieNode();
            
            cur = cur->children[idx];
        }
        
        cur->is_end = true;
    }
    
private:
    TrieNode* root;
};

class Solution {
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        Trie* trie = new Trie(words);
        TrieNode* root = trie->getRoot();
        set<string> res_set;
        
        for (int x=0; x < board.size(); ++x)
            for (int y=0; y < board[0].size(); ++y)
                findWords(board, x, y, root, "", res_set);
        
        // convert set to vector
        vector<string> res;
        for (auto it : res_set)
            res.push_back(it);
        
        return res;
    }

private:
    void findWords(vector<vector<char>>& board, int x, int y, TrieNode* root, string word, set<string>& result) {
        if (x < 0 || x >= board.size() || y < 0 || y >= board[0].size() || board[x][y]==' ')
            return;
        
        if (root->children[board[x][y] - 'a'] != NULL) {
            word = word + board[x][y];
            
            root = root->children[board[x][y] - 'a'];
            
            if (root->is_end)
                result.insert(word);
            
            char c = board[x][y];
            board[x][y] = ' ';
            
            findWords(board, x+1, y, root, word, result);
            findWords(board, x-1, y, root, word, result);
            findWords(board, x, y+1, root, word, result);
            findWords(board, x, y-1, root, word, result);
            board[x][y] = c;
        }
    }
};

File: 1. Problems/4. Trees/3_Word_Search_II.py
"""
LC 212 Word Search II

Given an m x n board of characters and a list of strings words, return all words on the board.

Each word must be constructed from letters of sequentially adjacent cells, where adjacent cells are horizontally or vertically neighboring. The same letter cell may not be used more than once in a word.

Example 1:

Input: board = [["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]], words = ["oath","pea","eat","rain"]
Output: ["eat","oath"]
Example 2:

Input: board = [["a","b"],["c","d"]], words = ["abcb"]
Output: []

Constraints:

m == board.length
n == board[i].length
1 <= m, n <= 12
board[i][j] is a lowercase English letter.
1 <= words.length <= 3 * 104
1 <= words[i].length <= 10
words[i] consists of lowercase English letters.
All the strings of words are unique.
"""
# Trie solution
class Solution:
    class TrieNode:
        def __init__(self):
            self.isWord = False
            self.next = {}
            
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        
        self.root = self.TrieNode()
        
        for word in words:
            node = self.root
            
            for char in word:
                if char not in node.next:
                    node.next[char] = self.TrieNode()
                    
                node = node.next[char]
                
            node.isWord = True
            
        self.res = set()
        self.dir = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        
        for i in range(len(board)):
            for j in range(len(board[0])):
                startNode = self.root
                self.searchWord(i, j, startNode, "", set(), board)
        
        return list(self.res)
        
    def searchWord(self, r: int, c: int, node: TreeNode, word: str, path: set, board: List[List[str]]):
        
        if r < 0 or r >= len(board) or c < 0 or c >= len(board[0]):
            return
        
        if board[r][c] not in node.next:
            return
        
        if (r, c) in path:
            return
        
        path.add((r, c))
        node = node.next[board[r][c]]
        
        if node.isWord:
            self.res.add(word + board[r][c])
            
        for dr, dc in self.dir:
            
            self.searchWord(r + dr, c + dc, node, word + board[r][c], path, board)
        
        path.remove((r, c))